NIFI-4325 Added new processor that uses the JSON DSL.

MikeThomsen opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,459 @@
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@EventDriven
+@SupportsBatching
+@Tags({"elasticsearch", "elasticsearch 5", "fetch", "read", "get"})
```
The tags should probably include "query" instead of "fetch" as does QueryElasticsearchHttp. Not sure whether they should include "json" or not, I'm fine with that either way :)

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,459 @@
+@SupportsBatching
+@Tags({"elasticsearch", "elasticsearch 5", "fetch", "read", "get"})
+@CapabilityDescription("Retrieves a document from Elasticsearch using the specified connection properties and the "
+        + "identifier of the document to retrieve. If the cluster has been configured for authorization and/or secure "
```
This processor does more than retrieve based on the identifier, it's based on the query right? Maybe elaborate on that as well as the various results and their relationships?

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,459 @@
+            .displayName("HTTP Hosts")
+            .description("A comma-separated list of HTTP hosts that host ElasticSearch query nodes.")
+            .required(true)
+            .expressionLanguageSupported(false)
```
Whether or not Expression Language has been historically supported in other ES processors, this seems like a good place to support them :) Pretty much goes for all these properties, except maybe sensitive properties (for now) like password. I realize a parameterized username might be useless without a parameterized password, but no harm no foul!

mattyb149 replied:
I think other processors have allowed EL in the password, it's not very secure without encrypted properties/registry entries, but that might be acceptable for some use cases

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,459 @@
+
+
+    @OnUnscheduled
+    public void onStartup() {
```
Just a style note, but this might be a misnamed/confusing method for an @OnUnscheduled annotation

MikeThomsen commented:
Thanks. I'll get to these some time today or tomorrow.

FYI Elastic has some official statements on the future direction of the Java client here: https://www.elastic.co/blog/state-of-the-official-elasticsearch-java-clients

The other processors will need reviewing to make sure they look good.

MikeThomsen commented:
I discovered a slick plugin for Maven that lets you cleanly run ElasticSearch in the background during integration tests. I added that to the test setup under a profile. So I have some cleanup to do on this branch to bring that in. I think it'll be worth it.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,451 @@
+@Tags({"elasticsearch", "elasticsearch 5", "query", "read", "get", "json"})
+@CapabilityDescription("A processor that allows the user to run a query (with aggregations) written with the " +
+        "ElasticSearch JSON DSL. It currently does not support pagination.")
+public class JsonQueryElasticsearch5 extends AbstractProcessor {
```
Since the newer Java client is just a wrapper around the REST API, that should make it pretty robust to work against many ES 5 clusters, right? We've been telling folks that the other ES5 processors only support 5.0.x clusters due to the native nature of the transport client, and the ESHttp processors are right for other cluster versions. If this new processor should work against many ES5 clusters, and is using the REST API under the hood, I wonder if we should call it JsonQueryElasticsearch5Http for consistency, what do you think?

MikeThomsen replied:
Yeah, that's worth considering.

mattyb149 replied:
Having the 5 and Http can be confusing as well, so I'm good with the name as-is :)

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,451 @@
+                if (sslService != null) {
+                    try {
+                        KeyStore keyStore = KeyStore.getInstance(sslService.getKeyStoreType());
+                        KeyStore trustStore = KeyStore.getInstance("JKS");
```
I think we should be using sslService.getTrustStoreType() here, in case it is set to something other than JKS

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,451 @@
+            Response resp = runQuery(query, index, type);
+
+            InputStream content = resp.getEntity().getContent();
+            String rawJson = IOUtils.toString(content, "UTF-8");
```
Is the response guaranteed to be in UTF-8 format?

MikeThomsen replied:
Apparently Lucene requires it: https://stackoverflow.com/a/23044215

mattyb149 commented:
If you're looking to run integration tests against a ES instance (from the Maven plugin), perhaps you can split out the integration tests into a new file called ITJsonQueryElasticsearch5, rather than having the @Ignore or running integration tests during the unit test phase. I realize the others aren't done this way, but if that Maven plugin works well, I think we should consider a separate Jira/PR to refactor the tests into proper integration vs unit tests, so they can be run during their corresponding lifecycle (integration tests aren't run during a normal test phase, e.g.)

MikeThomsen commented:
I set up a profile called integration-tests and renamed the file to JsonQueryElasticsearch_IT. It runs really well from the project folder, but for some reason running `mvn clean install -Pintegration-tests -pl :nifi-elasticsearch-5-processors` from the root folder causes surefire to be run twice (the first time passing, the second time failing because it stops ElasticSearch in between the two runs). When run from nifi-elasticsearch-5-processors it executes as expected.

MikeThomsen commented:
@mattyb149 I'm going to leave this open, but I decided to refactor the heck out of it around a client service for ElasticSearch. The service only has one method for now, but I think it's the way to go so that in the future as services become injectable in scripts and such, it'll be more flexible.

MikeThomsen commented:
@mattyb149 Ok. Changes are made. I refactored the commit to be based on a controller service. For now, that service only handles a single function: basic search. However, there is now a (hopefully) extensible base via a controller service to move forward. The controller service has integration tests only; the processor has junit tests with a mock service.

MikeThomsen commented:
@mattyb149 Have you had a chance to take a look at it?

joewitt commented:
@MikeThomsen It is awesome that you're contributing code/PRs.  As you can see though we do have quite a large backlog of PRs that need review/feedback/cycles.  Some of them, like integration with Mongo involve more effort than others to help test and review.  Please do consider helping by participating in/conducting reviews of others contributions as well.  That might help things like this be evaluated faster as there would be less overall backlog.  Thanks

MikeThomsen commented:
@joewitt alright, I'll see what trouble I can get into.

mattyb149 commented:
Reviewing...

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+            .name("el5-type")
+            .displayName("Type")
+            .description("The type of this document (used by Elasticsearch for indexing and searching)")
+            .defaultValue("")
```
If this is not a required field, you can remove the defaultValue() and replace the VALID validator with StandardValidators.NON_EMPTY_VALIDATOR. Not a requirement, just a suggestion to make it more consistent with other properties like this.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+    public static final PropertyDescriptor QUERY = new PropertyDescriptor.Builder()
+            .name("el5-query")
+            .displayName("Query")
+            .description("A query in JSON syntax, not Lucene syntax.")
```
An example would be really helpful here, although I realize it might look messy to have a JSON doc in the description. Any super-small query that would illustrate the kind of syntax to be used? If this is not the right place, perhaps in the CapabilityDescription or you could add an Additional Details page with some examples of queries, aggregations, etc.?

MikeThomsen replied:
Yeah, good point.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+
+    public static final PropertyDescriptor BREAK_UP_HITS = new PropertyDescriptor.Builder()
+            .name("el5-break-up-hits")
+            .displayName("Break up search results")
```
This is just a style comment (so not a requirement just a suggestion), other processors use the term "split" rather than "break up".

MikeThomsen replied:
Fair enough.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+            }
+        });
+        if (name != null) {
+            aggFlowFile = session.putAttribute(aggFlowFile, "aggregation.name", name);
```
Recommend adding a @WritesAttributes annotation to the processor class containing any attributes like aggregation.name, mime.type, etc. that may be added/updated by this processor

mattyb149 commented on nifi-nar-bundles/nifi-standard-services/nifi-elasticsearch-client-service-bundle/nifi-elasticsearch-client-service/pom.xml:
```diff
@@ -0,0 +1,158 @@
+        </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-hadoop-utils</artifactId>
```
Does this package need nifi-hadoop-utils? Please remove if not.

mattyb149 commented on nifi-nar-bundles/nifi-standard-services/nifi-elasticsearch-client-service-api/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientService.java:
```diff
@@ -0,0 +1,92 @@
+
+@Tags({"elasticsearch", "client"})
+@CapabilityDescription("A controller service for accessing an ElasticSearch client.")
+public interface ElasticSearchClientService extends ControllerService {
```
The current Http versions of the ES processors have the future capability of configuring a proxy, although these properties are not yet exposed to the user (see [NIFI-4198](https://issues.apache.org/jira/browse/NIFI-4198) and #2094); consider adding proxy capabilities here, or creating an improvement Jira after this PR/Jira has been merged.

MikeThomsen commented:
Thanks @mattyb149 I'll try to find some time to address these today or tomorrow. I've been knee deep in some Mongo-related work for a client. I have two pull requests for Mongo and record api-related fixes/updates if you're up for reviewing two small patches (one is 3 lines of code :) )

MikeThomsen commented on nifi-assembly/pom.xml:
```diff
@@ -338,6 +338,16 @@
+            <artifactId>nifi-elasticsearch-client-service-nar</artifactId>
+            <type>nar</type>
+        </dependency>
+<!--        <dependency>
```
That's how I included the nar file for the client service that the new processor uses. Is there a better approach?

mattyb149 replied:
I mean the commented out dependency, looks like dead code and/or you moved it to the processor's POM?

MikeThomsen commented:
@mattyb149 I think everything except the proxy issue is now in there.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,301 @@
+import java.util.Set;
+
+@WritesAttributes({
+    @WritesAttribute(attribute = "mime.type"),
```
This should have a description that the mime.type will be set to "application/json"

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,301 @@
+            if (hitsFlowFiles.size() > 0) {
+                session.transfer(hitsFlowFiles, REL_HITS);
+                for (FlowFile ff : hitsFlowFiles) {
+                    session.getProvenanceReporter().create(ff);
```
Since these are coming from an external system, I believe the convention is to use receive() instead of create(). Same goes below for aggregations.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,301 @@
+            .name("el5-query")
+            .displayName("Query")
+            .description("A query in JSON syntax, not Lucene syntax. Ex: " +
+                    "{\n" +
```
I haven't yet tried this out, do newlines and tabs show up in the tooltips ok?

MikeThomsen replied:
Nope.

mattyb149 replied:
No worries, just having an example (whether pretty-printed or not) is helpful

MikeThomsen commented:
@mattyb149 I think everything should be up to date now.

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.io.OutputStreamCallback;
```
Lines 38 and 42 are unused imports (i.e. CheckStyle violations), but I can fix these on final merge if there are no other issues :)

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+    @WritesAttribute(attribute = "mime.type", description = "application/json"),
+    @WritesAttribute(attribute = "aggregation.name", description = "The name of the aggregation whose results are in the output flowfile")
+})
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
```
It looks like you're not using the incoming flow file contents, in which case we might relax this constraint to INPUT_ALLOWED, and if no flow file is available (you can evaluate EL against a null FF) you can create one, what do you think?

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+    }
+
+
+    private ObjectMapper mapper = new ObjectMapper();
```
Nitpick, could be final?

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
```
Please see my other comment on allowing optional input flow files :)

mattyb149 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-5-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch5.java:
```diff
@@ -0,0 +1,291 @@
+            String type = context.getProperty(TYPE).evaluateAttributeExpressions(flowFile).getValue();
+
+            Optional<ElasticSearchResponse> resp = clientService.search(query, index, type);
+            ElasticSearchResponse response = resp.get();
```
the search() method returns an Optional, any chance the response object might not be present? Perhaps check isPresent() and send any incoming (not created) flow file to failure?

mattyb149 commented on nifi-nar-bundles/nifi-standard-services/nifi-elasticsearch-client-service-api/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchResponse.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public String toString() {
+        return "ElasticSearchResponse{" +
```
Nice!

mattyb149 commented on nifi-nar-bundles/nifi-standard-services/nifi-elasticsearch-client-service-bundle/nifi-elasticsearch-client-service-nar/src/main/resources/META-INF/NOTICE:
```diff
@@ -0,0 +1,332 @@
+nifi-hbase_1_1_2-client-service-nar
```
This appears to be a copy from a different NAR ;) The first line should be updated, and the rest of the NOTICES below should only be included if we are including the component(s) in the artifacts.

mattyb149 commented on nifi-nar-bundles/nifi-standard-services/nifi-elasticsearch-client-service-bundle/nifi-elasticsearch-client-service-nar/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,357 @@
+subcomponents is subject to the terms and conditions of the following
+licenses. 
+
+The binary distribution of this product bundles 'Jcodings' under an MIT style
```
The rest of this file should be inspected / updated depending on which components are actually included in the artifact(s). Please apply this to any NAR/module that has a LICENSE, and see my other comment on NOTICEs, thanks!

MikeThomsen commented:
@mattyb149  I think I got your changes done now. I renamed a few classes to remove obvious references to ElasticSearch 5.X because the goal is to make it an obvious fit with 6.X too now that that is out.

MikeThomsen commented:
@mattyb149 I've been thinking about what to do with this PR, and I think it needs to be put into a new NAR. I'm thinking "nifi-elasticsearch-restapi.nar" or something like that. What I'd ideally like to do is position this to be a new package that does full CRUD using only the official REST API from Elastic Co.

It might also help with the issue that @joewitt raised about the sizing because the current mix of ElasticSearch processors use different toolkits which needlessly raises the size of the NARs.

Thoughts?

mattyb149 commented:
Makes sense to me, we had to do the same with Elasticsearch 2.x and 5.x because of the difference in transport client versions. As folks move exclusively to ES5+, it would be nice to deprecate the other NARs and go forward with versions of all the processors (Get,Put,Search, etc.) using the new Java client.

MikeThomsen commented:
@mattyb149 I moved it to a new NAR and rebased against the latest master last night. So once this is committed, I'll start working on the new processors and service methods.

MikeThomsen commented:
@mattyb149 I think all of the change requests are done.

JPercivall commented:
Hey @mattyb149, what's your status for this review? From a cursory look, it appears like it just needs an updated pom from @MikeThomsen and the final approval. With talks of 1.6.0 happening it would be nice to get this in so those using ES 6 aren't limited to the HTTP processor. 

If help is needed to finalize things just let me know where I can help.

mattyb149 commented:
I'm not sure I'll have time to close the loop before 1.6.0, so if you'd like to finish the review/merge after Mike's rebase that would be very cool, thanks!

MikeThomsen commented:
Rebased and building. @mattyb149 I changed the functionality to match the GetMongo functionality you reviewed recently. Had to spend a while getting the build to work again.

JPercivall commented:
Reviewing

JPercivall commented:
Looks like the master build is failing with some check style issues, hence why your builds failed. I'll work on fixing them.

JPercivall commented on nifi-assembly/pom.xml:
```diff
@@ -445,6 +445,24 @@ language governing permissions and limitations under the License. -->
         </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-elasticsearch-client-service-api-nar</artifactId>
```
My build is failing due to this. I don't see this artifact created anywhere.

MikeThomsen replied:
Weird, I thought that I took that out.

JPercivall commented:
I'm seeing NPE and FlowFile handling exceptions. I didn't do anything special, just using the example query. I'll attach a template. Errors from the logs below.


> 2018-03-12 23:58:11,386 ERROR [Timer-Driven Process Thread-10] o.a.n.p.e.JsonQueryElasticsearch JsonQueryElasticsearch[id=1d5d83d1-0162-1000-cdf2-92fa9e61ef42] Error processing flowfile.: java.lang.NullPointerException
java.lang.NullPointerException: null
        at org.apache.nifi.processors.elasticsearch.JsonQueryElasticsearch.onTrigger(JsonQueryElasticsearch.java:248)
        at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27)
        at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1123)
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:147)
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47)
        at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:128)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
2018-03-12 23:58:11,389 ERROR [Timer-Driven Process Thread-10] o.a.n.p.e.JsonQueryElasticsearch JsonQueryElasticsearch[id=1d5d83d1-0162-1000-cdf2-92fa9e61ef42] JsonQueryElasticsearch[id=1d5d83d1-0162-1000-cdf2-92fa9e61ef42] failed to process due to org.apache.nifi.processor.exception.FlowFileHandlingException: StandardFlowFileRecord[uuid=77ad5fff-d4b6-4e11-8d8c-b07ac1d0a6cb,claim=StandardContentClaim [resourceClaim=StandardResourceClaim[id=1520913052434-1, container=default, section=1], offset=30, length=228],offset=226,name=46172028496405,size=2] transfer relationship not specified; rolling back session: {}
org.apache.nifi.processor.exception.FlowFileHandlingException: StandardFlowFileRecord[uuid=77ad5fff-d4b6-4e11-8d8c-b07ac1d0a6cb,claim=StandardContentClaim [resourceClaim=StandardResourceClaim[id=1520913052434-1, container=default, section=1], offset=30, length=228],offset=226,name=46172028496405,size=2] transfer relationship not specified
        at org.apache.nifi.controller.repository.StandardProcessSession.checkpoint(StandardProcessSession.java:251)
        at org.apache.nifi.controller.repository.StandardProcessSession.commit(StandardProcessSession.java:321)
        at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:28)
        at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1123)
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:147)
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47)
        at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:128)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

JPercivall commented:
Template for FlowFile handling exception.

[FlowFile_Handling_Exception.txt](https://github.com/apache/nifi/files/1805420/FlowFile_Handling_Exception.txt)

MikeThomsen commented:
@JPercivall @mattyb149 That exception only happens for me when I stop the processor. It happens here:

``
if (hitsFlowFiles.size() > 0) {
    session.transfer(hitsFlowFiles, REL_HITS);
    for (FlowFile ff : hitsFlowFiles) {
        session.getProvenanceReporter().send(ff, clientService.getTransitUrl(index, type));
    }
}
``

On the provenance reporting line. Any ideas?

MikeThomsen commented:
@JPercivall @mattyb149 Once I changed the code to use `transfer(FlowFile, Relationship)` and not `transfer(List, Relationship)` the error stopped happening.

MikeThomsen commented:
2/3 builds pass now.

MikeThomsen commented:
@JPercivall It's ready for review.

JPercivall commented:
Thanks @MikeThomsen I should be able to take another look later tonight

JPercivall commented:
The nifi-elasticsearch-client-service-nar is missing a license and notice.

JPercivall commented:
Apologises, won't be able to finish the review this evening. I reviewed the restapi nar License and Notice files they look good to go. Just need to do a bit of functional testing and verify @mattyb149 comments were addressed and we should be ready to merge (pending the L/N for the service nar).

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java:
```diff
@@ -0,0 +1,326 @@
+
+    @OnUnscheduled
+    public void onUnscheduled() {
+        this.clientService = null;
```
This should be changed from "OnUnscheduled"[1] to "OnStopped"[2]. "OnUnscheduled" will be run as soon as the user clicks stop. This stops the processor from being schedueld again but there could still be lingering tasks running (thus leading to null pointer exceptions when it tries to use the clientService). "OnStopped" is run once all tasks finish.

[1] https://github.com/apache/nifi/blob/master/nifi-api/src/main/java/org/apache/nifi/annotation/lifecycle/OnUnscheduled.java
[2] https://github.com/apache/nifi/blob/master/nifi-api/src/main/java/org/apache/nifi/annotation/lifecycle/OnStopped.java

MikeThomsen commented:
@JPercivall Done.

JPercivall commented:
Thanks @MikeThomsen, looks like that fixed the NPE. Still needs the License and Notice in the client service nar.

Also, it's a bit of an anti-pattern to have the logging and error handling in the Controller Service for methods called by the processor. This is because any time there is a problem with the search, the bulletin is logged at the Controller Service. This means that not only will a user not know which processor is the cause of the error (since it could be shared) but also when looking at the map the processor appears to be working without issue. Lastly, since the error isn't bubbled back to the processor, it never yields so if there is something wrong with the query it will continually error and log without pause. 

Related, there's also some weirdness with error handling when there is an incoming connection. Currently, input is only sent to Failure instead of Original when an error is caught at the top level. This means that if a query is messed up and is caught as a JsonProcessingException in the Controller service, the input will be routed to original and no other outputs. As a user, I would expect that flowfile to be sent to Failure.

So in summary, the CS should let the calling processor decide how to handle any exceptions encountered during processing. This allows the processor to more granularlly decide what to do under different situations. For example, was it a JSON parsing exception and it's acting as a source processor? well we better yield the whole processor since it's probably just going to continually fail. Is there an incoming connection but some network error happened when searching? again, later FlowFiles will probably fail too, so we should yield. If you have any questions on when to yield the processor vs penalizing the FlowFile, feel free to ask.

Lastly, it's preferred if you don't squash your commits every time. If you don't, that allows the reviewer to more easily see exactly what changed since they last reviewed it. Also allows reviewers to see how the PR evolved over time in response to different comments.

JPercivall commented on nifi-nar-bundles/nifi-standard-services/pom.xml:
```diff
@@ -15,6 +15,18 @@
 -->
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
+    <dependencies>
```
This dependencies section shouldn't be needed here. This is the parent pom for all of the standard services.

MikeThomsen replied:
Done.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java:
```diff
@@ -0,0 +1,326 @@
+                if (hitsFlowFiles.size() > 0) {
+                    for (FlowFile ff : hitsFlowFiles) {
+                        session.transfer(ff, REL_HITS);
+                        session.getProvenanceReporter().send(ff, clientService.getTransitUrl(index, type));
```
This could be improved a bit by setting a single String and re-using instead it of recalculating for every flowfile hit (and agg hit).

MikeThomsen replied:
Done.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java:
```diff
@@ -0,0 +1,326 @@
+
+                if (hitsFlowFiles.size() > 0) {
+                    for (FlowFile ff : hitsFlowFiles) {
+                        session.transfer(ff, REL_HITS);
```
This should be able to be switched by to the single transfer since the root issue was setting the clientService to null on unScheduled.

MikeThomsen replied:
Done.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java:
```diff
@@ -0,0 +1,326 @@
+    private static final Set<Relationship> relationships;
+    private static final List<PropertyDescriptor> propertyDescriptors;
+
+    private ElasticSearchClientService clientService;
```
This should be set as volatile since it is referenced by multiple different threads.

MikeThomsen replied:
Done.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -0,0 +1,255 @@
+public class ElasticSearchClientServiceImpl extends AbstractControllerService implements ElasticSearchClientService {
+    private ObjectMapper mapper = new ObjectMapper();
+
+    private List<PropertyDescriptor> properties;
```
Can this be declared similar to the properties in the processor and marked final?

MikeThomsen replied:
Done.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -0,0 +1,255 @@
+        this.url = null;
+    }
+
+    private void setupClient(ConfigurationContext context) throws Exception {
```
Try to avoid "throws Exception". Looks like the only one thrown is "MalformedURLException".

MikeThomsen replied:
Done.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -0,0 +1,255 @@
+                        return requestConfigBuilder;
+                    } catch (Exception ex) {
+                        try {
+                            PrintWriter writer = new PrintWriter(new FileWriter("/tmp/out.log"));
```
Why write to "/tmp/out.log"?

MikeThomsen replied:
Something I wrote to make it easier to dump the raw query. Shouldn't be in there.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -0,0 +1,255 @@
+                            context1.init(kmf.getKeyManagers(), tmf.getTrustManagers(), new SecureRandom());
+                            httpClientBuilder = httpClientBuilder.setSSLContext(context1);
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
```
Why throw RTE here? (and below)

MikeThomsen replied:
Removed. I think I have a better solution for that now.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java:
```diff
@@ -0,0 +1,326 @@
+@EventDriven
+@Tags({"elasticsearch", "elasticsearch 5", "query", "read", "get", "json"})
+@CapabilityDescription("A processor that allows the user to run a query (with aggregations) written with the " +
+        "ElasticSearch JSON DSL. It currently does not support pagination.")
```
Related to the note about pagination, should add something indicating that very large searches should be done as the response is loaded into memory.

MikeThomsen replied:
Ok. I should also note that scroll queries are not supported.

JPercivall replied:
I don't see this in the latest commit

MikeThomsen commented:
@JPercivall Thanks for the feedback. I'll get to working on these. WRT:

> Lastly, it's preferred if you don't squash your commits every time. If you don't, that allows the reviewer to more easily see exactly what changed since they last reviewed it. Also allows reviewers to see how the PR evolved over time in response to different comments.

Ok, I can do that from now on. Shouldn't they be squashed before a merge into master?

MikeThomsen commented:
Ok. It should all be there now.

JPercivall commented:
Taking a look now. WRT:

> Ok, I can do that from now on. Shouldn't they be squashed before a merge into master?

Yup, they will be squashed prior to merge to master but the reviewer can do that at the same time they add the comment to the commit to close the PR.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -121,7 +119,7 @@ private void setupClient(ConfigurationContext context) throws Exception {
                 .setHttpClientConfigCallback(httpClientBuilder -> {
-                    if (sslService != null) {
+                    if (sslService != null && sslService.isKeyStoreConfigured() && sslService.isTrustStoreConfigured()) {
                         try {
```
Given that this is a callback, we probably want to move as much of the logic out of here as possible. That way we can better handle any errors and reduce performance cost if this is called multiple times. Currently, if there is an error which hits line 145, the error will be caught and logged but it will continue to create the http client as if nothing went wrong. What we should probably do is have a final SSLContext which gets created above, bubbling up any errors so that the CS doesn't enable in a bad state, and set the SSLContext same way as below in the context.

MikeThomsen replied:
Done. I moved that logic out and have the it bubble it up with an InitializationException if any of those exceptions are thrown.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -61,22 +58,23 @@
-        properties.add(ElasticSearchClientService.SOCKET_TIMEOUT);
-        properties.add(ElasticSearchClientService.RETRY_TIMEOUT);
+    static {
+        List _props = new ArrayList();
```
Bit of a nit pick but can this init be changed so it's not of raw type:

> List\<PropertyDescriptor\> _props = new ArrayList\<\>();

MikeThomsen replied:
Done.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java:
```diff
@@ -0,0 +1,329 @@
+            retVal = context.getProperty(QUERY).evaluateAttributeExpressions(input).getValue();
+        } else if (input != null) {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            session.exportTo(input, out);
```
Looks like if the Query property isn't set and there is an incoming FF then it will use the FF's content as the query. I don't see this documented in the Query props. This should at least be documented and also ideally, there would be some custom validation too.

MikeThomsen replied:
Yeah, it needs the documentation. That was a recent addition. I'm not sure about the validation because it's expecting a JSON payload, and if that's malformed in any respect the client service will throw an exception back to the processor now.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/JsonQueryElasticsearch.java:
```diff
@@ -0,0 +1,329 @@
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor QUERY_ATTRIBUTE = new PropertyDescriptor.Builder()
```
Where is this prop used? I only see it referenced when it's added to the descriptors. I do see the name named attribute used a lot in the mongo processors though.

MikeThomsen replied:
Fixed.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -0,0 +1,225 @@
+            getLogger().debug(sb.toString());
+        }
+
+        return Optional.of(esr);
```
Does this still need to be an Optional?

MikeThomsen replied:
Probably not.

JPercivall commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -0,0 +1,225 @@
+    @Override
+    public Optional<SearchResponse> search(String query, String index, String type) throws IOException {
+        Response response = runQuery(query, index, type);
+        Map<String, Object> parsed = mapper.readValue(IOUtils.toString(response.getEntity().getContent()), Map.class);
```
2 things:
1. IOUtils.toString is deprecated, you should pass the expected charset to use.
2. In general, you should avoid doing a lot of nesting of logic into a single line. It helps a lot with readability and diagnosing errors. For example, if we hit an NPE here, it could be for any of 3 different objects. At the very least, the IOUtils call to get the response content as a string should be on another line.

JPercivall replied:
Also, prior to interpreting this as a map, the status line of the response should be checked to see if it actually succeeded. That way we can log the ES error message in the processor as a bulletin for easy diagnosing of problems by the user. You can see handling of the status message in the FetchElasticsearcHttp processor: https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearchHttp.java#L262

JPercivall commented:
Hey @MikeThomsen, just finished up another round of reviewing. Again, mostly focused on the error handling portion. 

I hope my review doesn't come off in the wrong way, there is a lot of great work here and I just want to make sure the usability is top notch. Error handling is a huge focus in NiFi because we're writing up a toolbox of processors for potentially non-technical users to work with however they want. Those users need to be able to see what they messed up and what went wrong for efficient and effective rectification. 

Thanks for the hard work, and the willingness to iterate and learn!

MikeThomsen commented:
> I hope my review doesn't come off in the wrong way, there is a lot of great work here and I just want to make sure the usability is top notch.

Not at all. It's all fair and good feedback. I'll take a crack at these tomorrow.

MikeThomsen commented:
@JPercivall changes checked in.

MikeThomsen commented:
@JPercivall Any chance we can close this out?

JPercivall commented:
Hey @MikeThomsen, I'm planning on reviewing this tomorrow evening

MikeThomsen commented:
Thanks.

MikeThomsen commented:
@JPercivall FYI, part of the reason I want to get this done is I'm planning on doing a whole new set of ES processors that are based on bring the total CRUD functionality over to the official Elastic-provided APIs. The current set of processors use the transport API (deprecated) and make manual REST calls that AFAIK don't do master detection and stuff like that which comes baked-in with the Elastic APIs.

JPercivall commented:
Hey @MikeThomsen, just reviewed the changes. They look good and I think we're almost there. One thing that is missing is the comment regarding pulling the results into memory.

Also, I tried to build NiFi and test the processor an instance but hit the following error when I attempted to start up NiFi (./bin/nifi.sh start). I tried rebasing it to the latest master but to no avail. 

> 2018-03-24 02:06:53,171 ERROR [main] org.apache.nifi.NiFi Failure to launch NiFi due to java.util.ServiceConfigurationError: org.apache.nifi.processor.Processor: Provider org.apache.nifi.processors.elasticsearch.JsonQueryElasticsearch could not be instantiated
java.util.ServiceConfigurationError: org.apache.nifi.processor.Processor: Provider org.apache.nifi.processors.elasticsearch.JsonQueryElasticsearch could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.apache.nifi.nar.ExtensionManager.loadExtensions(ExtensionManager.java:148)
	at org.apache.nifi.nar.ExtensionManager.discoverExtensions(ExtensionManager.java:123)
	at org.apache.nifi.web.server.JettyServer.start(JettyServer.java:771)
	at org.apache.nifi.NiFi.<init>(NiFi.java:157)
	at org.apache.nifi.NiFi.<init>(NiFi.java:71)
	at org.apache.nifi.NiFi.main(NiFi.java:292)
Caused by: java.lang.NoClassDefFoundError: com/fasterxml/jackson/databind/ObjectMapper
	at org.apache.nifi.processors.elasticsearch.JsonQueryElasticsearch.<init>(JsonQueryElasticsearch.java:206)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 8 common frames omitted
Caused by: java.lang.ClassNotFoundException: com.fasterxml.jackson.databind.ObjectMapper
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 15 common frames omitted

MikeThomsen commented:
@JPercivall I fixed the dependency issue and tried a simple flow. Everything seemed to work.

JPercivall commented:
Looks good, was able to do a couple simple queries and aggregations. I wasn't able to test HTTPS though as I don't have access to the X-Pack. Were you able to do so @MikeThomsen?

The only lingering issue, that I've mentioned multiple times now, is to add a note about the entire body of the response being loaded into memory.

MikeThomsen commented:
@JPercivall I just got done testing it w/ X-Pack for SSL and u/p auth based on Elastic docs. Everything was in working order there. [Steps if you want to validate](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/configuring-tls.html)

I also added the documentation there. Sorry about that.

AFAIK, that warning is going to be more of a polite warning than a real "here be dragons" sort of thing because IIRC ES limits you to 10k results unless you explicitly up the number in the index configuration. Probably a good reminder, but I doubt anyone is going to blow up their NiFi installation without having to take deliberate steps to facilitate that.

JPercivall commented:
+1 

Overall looks good and works as expected. I've built with contrib check and tested using a vanilla instance. I will squash and merge to master. Thanks @MikeThomsen.
